csrc/scheduler/tools/resize_utils.h

```diff
@@ -19,5 +22,81 @@ namespace scheduler_tools {
 // fusion inputs are skipped as their loop domains don't matter.
 void propagateResizeToInputs(Expr* resize_op);
 
+// Given a topologically ordered list of resize-based tensor ops such
+// as slice and pad, check if they can be propagated to fusion inputs
+// exclusively without causing any visible side effect. For example,
+// if a tensor is sliced and also is used to produce an output without
+// the slicing, the slice is considered non exclusive as the slice
+// input has the other visible consumer. Propagating the resize of the
+// slice to the slice input is invalid since the output computed from
+// the slice input depends on the full iteration space.
+//
+// For example, consider the following case:
+//
+// t0 = makeSymbolicTensor(1)
+// fusion.addInput(t0)
+// t1 = t0 + 1
+// t2 = t1[1:10]
+// t3 = t1 + 1
+// fusion.addOutput(t2)
+// fusion.addOutput(t3)
+//
+// In this case, propating the resize op of the slice would alter t1,
+// which would in turn affect t3, which is a fusion output. Since the
+// change would be visible due to the change of t3, this resie op is
+// considered non-exclusive.
+//
+// Consider a slightly different case as shown below:
+//
+// t0 = makeSymbolicTensor(1)
+// fusion.addInput(t0)
+// t1 = t0[1:10]
+// t2 = t0 + 1
+// fusion.addOutput(t1)
+// fusion.addOutput(t2)
+//
+// Note that the slice is directly done with the fusion input. Since
+// we do not propagate resize ops to fusion inputs, this can be
+// considered exclusive. However, this is also considered
+// non-exclusive since the actual scheduling inserts a cache after t0,
+// which can cause a visible side effect if the resize is propagated.
+//
+// Another non-exclusivess comes from dependent fusion outputs. For
+// example, if a slice input depends on a fusion output, propation
+// would alter the fusion output. Consider a case like:
+//
+// t0 = makeSymbolicTensor(1)
+// fusion.addInput(t0)
+// t1 = t0 + 1
+// t2 = t1[1:10] // slice
+// fusion.addOutput(t1)
+// fusion.addOutput(t2)
+//
+// If the resize op for the slice is propagated to t1, only the
+// section of [1:10] would be computed. Since that would change a
+// fusion output, the resize op is considered non-exclusive.
+//
+// When there's a chain of resize-based ops, for example:
+//
+// t0 = makeSymbolicTensor(1)
+// fusion.addInput(t0)
+// t1 = t0 + 1
+// t2 = t1[1:10]
+// t3 = t2[2:5]
+// t4 = t1 + 1
+// fusion.addOutput(t3)
+// fusion.addOutput(t4)
+//
+// We do not consider the second slice as non-exclusive as
+// long as the first slice is considered non-exclusive. This will be
```

Comment:
Is this a typo? 
```suggestion
// We do not consider the second slice as exclusive as
// long as the first slice is considered non-exclusive. This will be
```

Reply:
Yes, you're right

Reply:
Ah, actually, it is not a typo. It doesn't matter at this PR, but in the next PR, I'm going to resolve these non-exclusive resize ops by replication. Here, we don't want to consider the second slice non-exclusive because once the first slice is changed to be exclusive, the second slice should have no problem of propagating its resize op to forward. More specifically, we would replicate `t1` for the first slice because the first slice is not exclusive, but that's enough to make the fusion work with the scheduler.

After the replication, it would look like:

```
t0 = makeSymbolicTensor(1)
fusion.addInput(t0)
t1 = t0 + 1
t1_copy = t0 + 1
t2 = t1_copy[1:10]
t3 = t2[2:5]
t4 = t1 + 1
fusion.addOutput(t3)
fusion.addOutput(t4)
```

Then, `propagateResizeToInputs` would first propagate the resize of the first slice from `t2` to `t1_copy`. We don't update fusion inputs, so that's it. Then, `propagateResizeToInputs` would be done for the second slice, from `t3` to `t2`, then `t1_copy`. This propagation would not affect `t4`, so it should be safe to do.

Reply:
Ah ok. Thank you for the explanation.